macOS M2 32 GB -- processing failed

I get error message:
"Error: llama runner process has terminated"

Does that mean it run out of memory?

Is it possible to make it smaller?

Same issue with m1 pro 32GB
In my situation, I use llm : dolphin-2.5-mixtral-8x7b
related logs:
```
[GIN] 2023/12/16 - 22:24:41 | 200 |      63.875µs |       127.0.0.1 | HEAD     "/"
[GIN] 2023/12/16 - 22:24:41 | 200 |    4.554292ms |       127.0.0.1 | POST     "/api/show"
[GIN] 2023/12/16 - 22:24:42 | 200 |     862.208µs |       127.0.0.1 | POST     "/api/show"
2023/12/16 22:24:42 llama.go:436: starting llama runner
2023/12/16 22:24:42 llama.go:494: waiting for llama runner to start responding
{"timestamp":1702736682,"level":"INFO","function":"main","line":2653,"message":"build info","build":441,"commit":"948ff13"}
{"timestamp":1702736682,"level":"INFO","function":"main","line":2660,"message":"system info","n_threads":8,"n_threads_batch":-1,"total_threads":10,"system_info":"AVX = 0 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | SSSE3 = 0 | VSX = 0 | "}
llama_model_loader: loaded meta data with 23 key-value pairs and 995 tensors from /Users/fengji/.ollama/models/blobs/sha256:bdb11b0699e03d791f0accd97279989d810d79615c6cf5ac21fb68e8f33e8ca3 (version GGUF V3 (latest))
llama_model_loader: - tensor    0:                token_embd.weight q4_0     [  4096, 32002,     1,     1 ]
llama_model_loader: - tensor    1:              blk.0.attn_q.weight q4_0     [  4096,  4096,     1,     1 ]
llama_model_loader: - tensor    2:              blk.0.attn_k.weight q8_0     [  4096,  1024,     1,     1 ]
llama_model_loader: - tensor    3:              blk.0.attn_v.weight q8_0     [  4096,  1024,     1,     1 ]
llama_model_loader: - tensor    4:         blk.0.attn_output.weight q4_0     [  4096,  4096,     1,     1 ]
llama_model_loader: - tensor    5:        blk.0.ffn_gate_inp.weight f16      [  4096,     8,     1,     1 ]
llama_model_loader: - tensor    6:          blk.0.ffn_gate.0.weight q4_0     [  4096, 14336,     1,     1 ]
llama_model_loader: - tensor    7:          blk.0.ffn_down.0.weight q4_0     [ 14336,  4096,     1,     1 ]
llama_model_loader: - tensor    8:            blk.0.ffn_up.0.weight q4_0     [  4096, 14336,     1,     1 ]
llama_model_loader: - tensor    9:          blk.0.ffn_gate.1.weight q4_0     [  4096, 14336,     1,     1 ]
llama_model_loader: - tensor   10:          blk.0.ffn_down.1.weight q4_0     [ 14336,  4096,     1,     1 ]
llama_model_loader: - tensor   11:            blk.0.ffn_up.1.weight q4_0     [  4096, 14336,     1,     1 ]
llama_model_loader: - tensor   12:          blk.0.ffn_gate.2.weight q4_0     [  4096, 14336,     1,     1 ]
llama_model_loader: - tensor   13:          blk.0.ffn_down.2.weight q4_0     [ 14336,  4096,     1,     1 ]
llama_model_loader: - tensor   14:            blk.0.ffn_up.2.weight q4_0     [  4096, 14336,     1,     1 ]
...
...
llama_model_loader: - tensor  993:               output_norm.weight f32      [  4096,     1,     1,     1 ]
llama_model_loader: - tensor  994:                    output.weight q6_K     [  4096, 32002,     1,     1 ]
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.name str              = ehartford
llama_model_loader: - kv   2:                       llama.context_length u32              = 32768
llama_model_loader: - kv   3:                     llama.embedding_length u32              = 4096
llama_model_loader: - kv   4:                          llama.block_count u32              = 32
llama_model_loader: - kv   5:                  llama.feed_forward_length u32              = 14336
llama_model_loader: - kv   6:                 llama.rope.dimension_count u32              = 128
llama_model_loader: - kv   7:                 llama.attention.head_count u32              = 32
llama_model_loader: - kv   8:              llama.attention.head_count_kv u32              = 8
llama_model_loader: - kv   9:                         llama.expert_count u32              = 8
llama_model_loader: - kv  10:                    llama.expert_used_count u32              = 2
llama_model_loader: - kv  11:     llama.attention.layer_norm_rms_epsilon f32              = 0.000010
llama_model_loader: - kv  12:                       llama.rope.freq_base f32              = 1000000.000000
llama_model_loader: - kv  13:                          general.file_type u32              = 2
llama_model_loader: - kv  14:                       tokenizer.ggml.model str              = llama
llama_model_loader: - kv  15:                      tokenizer.ggml.tokens arr[str,32002]   = ["<unk>", "<s>", "</s>", "<0x00>", "<...
llama_model_loader: - kv  16:                      tokenizer.ggml.scores arr[f32,32002]   = [0.000000, 0.000000, 0.000000, 0.0000...
llama_model_loader: - kv  17:                  tokenizer.ggml.token_type arr[i32,32002]   = [2, 3, 3, 6, 6, 6, 6, 6, 6, 6, 6, 6, ...
llama_model_loader: - kv  18:                tokenizer.ggml.bos_token_id u32              = 1
llama_model_loader: - kv  19:                tokenizer.ggml.eos_token_id u32              = 32000
llama_model_loader: - kv  20:               tokenizer.ggml.add_bos_token bool             = true
llama_model_loader: - kv  21:               tokenizer.ggml.add_eos_token bool             = false
llama_model_loader: - kv  22:               general.quantization_version u32              = 2
llama_model_loader: - type  f32:   65 tensors
llama_model_loader: - type  f16:   32 tensors
llama_model_loader: - type q4_0:  833 tensors
llama_model_loader: - type q8_0:   64 tensors
llama_model_loader: - type q6_K:    1 tensors
llm_load_vocab: special tokens definition check successful ( 261/32002 ).
llm_load_print_meta: format           = GGUF V3 (latest)
llm_load_print_meta: arch             = llama
llm_load_print_meta: vocab type       = SPM
llm_load_print_meta: n_vocab          = 32002
llm_load_print_meta: n_merges         = 0
llm_load_print_meta: n_ctx_train      = 32768
llm_load_print_meta: n_embd           = 4096
llm_load_print_meta: n_head           = 32
llm_load_print_meta: n_head_kv        = 8
llm_load_print_meta: n_layer          = 32
llm_load_print_meta: n_rot            = 128
llm_load_print_meta: n_gqa            = 4
llm_load_print_meta: f_norm_eps       = 0.0e+00
llm_load_print_meta: f_norm_rms_eps   = 1.0e-05
llm_load_print_meta: f_clamp_kqv      = 0.0e+00
llm_load_print_meta: f_max_alibi_bias = 0.0e+00
llm_load_print_meta: n_ff             = 14336
llm_load_print_meta: n_expert         = 8
llm_load_print_meta: n_expert_used    = 2
llm_load_print_meta: rope scaling     = linear
llm_load_print_meta: freq_base_train  = 1000000.0
llm_load_print_meta: freq_scale_train = 1
llm_load_print_meta: n_yarn_orig_ctx  = 32768
llm_load_print_meta: rope_finetuned   = unknown
llm_load_print_meta: model type       = 7B
llm_load_print_meta: model ftype      = mostly Q4_0
llm_load_print_meta: model params     = 46.70 B
llm_load_print_meta: model size       = 24.62 GiB (4.53 BPW)
llm_load_print_meta: general.name     = ehartford
llm_load_print_meta: BOS token        = 1 '<s>'
llm_load_print_meta: EOS token        = 32000 '<|im_end|>'
llm_load_print_meta: UNK token        = 0 '<unk>'
llm_load_print_meta: LF token         = 13 '<0x0A>'
llm_load_tensors: ggml ctx size =    0.39 MiB
llm_load_tensors: mem required  = 25216.27 MiB
.....
.......
...
....
....
...
..
...
.
..
..
.
..
.
.
..
.
.
.
.
.
.
.
.
.
.
.
.
.
...
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
llama_new_context_with_model: n_ctx      = 4096
llama_new_context_with_model: freq_base  = 1000000.0
llama_new_context_with_model: freq_scale = 1
llama_new_context_with_model: KV self size  =  512.00 MiB, K (f16):  256.00 MiB, V (f16):  256.00 MiB
llama_build_graph: non-view tensors processed: 1124/1124
ggml_metal_init: allocating
ggml_metal_init: found device: Apple M1 Pro
ggml_metal_init: picking default device: Apple M1 Pro
ggml_metal_init: default.metallib not found, loading from source
ggml_metal_init: GGML_METAL_PATH_RESOURCES = nil
ggml_metal_init: loading '/var/folders/jd/_9xbz5493rn_m477kjkvp5_m0000gn/T/ollama2957943658/llama.cpp/gguf/build/metal/bin/ggml-metal.metal'
ggml_metal_init: GPU name:   Apple M1 Pro
ggml_metal_init: GPU family: MTLGPUFamilyApple7 (1007)
ggml_metal_init: hasUnifiedMemory              = true
ggml_metal_init: recommendedMaxWorkingSetSize  = 22906.50 MB
ggml_metal_init: maxTransferRate               = built-in GPU
llama_new_context_with_model: compute buffer total size = 319.35 MiB
llama_new_context_with_model: max tensor size =   102.55 MiB
ggml_metal_add_buffer: allocated 'data            ' buffer, size = 16384.00 MiB, offs =            0
ggml_metal_add_buffer: allocated 'data            ' buffer, size =  8935.19 MiB, offs =  17072324608, (25319.88 / 21845.34)ggml_metal_add_buffer: warning: current allocated size is greater than the recommended max working set size
ggml_metal_add_buffer: allocated 'kv              ' buffer, size =   512.03 MiB, (25831.91 / 21845.34)ggml_metal_add_buffer: warning: current allocated size is greater than the recommended max working set size
ggml_metal_add_buffer: allocated 'alloc           ' buffer, size =   316.05 MiB, (26147.95 / 21845.34)ggml_metal_add_buffer: warning: current allocated size is greater than the recommended max working set size
ggml_metal_graph_compute: command buffer 4 failed with status 5
GGML_ASSERT: /Users/jmorgan/workspace/ollama/llm/llama.cpp/gguf/ggml-metal.m:2353: false
2023/12/16 22:24:54 llama.go:451: signal: abort trap
2023/12/16 22:24:54 llama.go:459: error starting llama runner: llama runner process has terminated
2023/12/16 22:24:54 llama.go:525: llama runner stopped successfully
[GIN] 2023/12/16 - 22:24:54 | 500 | 12.347520041s |       127.0.0.1 | POST     "/api/generate"
```

--- a/gpu/gpu.go
+++ b/gpu/gpu.go
@@ -132,7 +132,7 @@ func CheckVRAM() (int64, error) {
 	gpuInfo := GetGPUInfo()
 	if gpuInfo.FreeMemory > 0 && (gpuInfo.Library == "cuda" || gpuInfo.Library == "rocm") {
 		// allocate 384MiB for llama.cpp overhead (outside of model)
-		overhead := 384 * 1024 * 1024
+		overhead := uint64(384 * 1024 * 1024)
 		if gpuInfo.FreeMemory <= overhead {
 			return 0, nil
 		}

--- a/gpu/gpu.go
+++ b/gpu/gpu.go
@@ -131,7 +131,14 @@ func getCPUMem() (memInfo, error) {
 func CheckVRAM() (int64, error) {
 	gpuInfo := GetGPUInfo()
 	if gpuInfo.FreeMemory > 0 && (gpuInfo.Library == "cuda" || gpuInfo.Library == "rocm") {
-		return int64(gpuInfo.FreeMemory), nil
+		// allocate 384MiB for llama.cpp overhead (outside of model)
+		overhead := 384 * 1024 * 1024
+		if gpuInfo.FreeMemory <= overhead {
+			return 0, nil
+		}
+
+		return int64(gpuInfo.FreeMemory - overhead), nil
 	}
+
 	return 0, fmt.Errorf("no GPU detected") // TODO - better handling of CPU based memory determiniation
 }

--- a/gpu/gpu.go
+++ b/gpu/gpu.go
@@ -16,8 +16,6 @@ import (
 	"runtime"
 	"sync"
 	"unsafe"
-
-	"github.com/jmorganca/ollama/api"
 )
 
 type handles struct {
@@ -136,28 +134,4 @@ func CheckVRAM() (int64, error) {
 		return int64(gpuInfo.FreeMemory), nil
 	}
 	return 0, fmt.Errorf("no GPU detected") // TODO - better handling of CPU based memory determiniation
-}
-
-func NumGPU(numLayer, fileSizeBytes int64, opts api.Options) int {
-	if opts.NumGPU != -1 {
-		return opts.NumGPU
-	}
-	info := GetGPUInfo()
-	if info.Library == "cpu" || info.Library == "default" {
-		return 0
-	}
-
-	/*
-		Calculate bytes per layer, this will roughly be the size of the model file divided by the number of layers.
-		We can store the model weights and the kv cache in vram,
-		to enable kv chache vram storage add two additional layers to the number of layers retrieved from the model file.
-	*/
-	bytesPerLayer := uint64(fileSizeBytes / numLayer)
-
-	// 75% of the absolute max number of layers we can fit in available VRAM, off-loading too many layers to the GPU can cause OOM errors
-	layers := int(info.FreeMemory/bytesPerLayer) * 3 / 4
-
-	log.Printf("%d MB VRAM available, loading up to %d %s GPU layers out of %d", info.FreeMemory/(1024*1024), layers, info.Library, numLayer)
-
-	return layers
 }

--- a/gpu/gpu_darwin.go
+++ b/gpu/gpu_darwin.go
@@ -6,18 +6,30 @@ import "C"
 import (
 	"runtime"
 
-	"github.com/jmorganca/ollama/api"
+	"github.com/pbnjay/memory"
 )
 
 // CheckVRAM returns the free VRAM in bytes on Linux machines with NVIDIA GPUs
 func CheckVRAM() (int64, error) {
-	// TODO - assume metal, and return free memory?
-	return 0, nil
+	if runtime.GOARCH == "amd64" {
+		return 0, nil
+	}
 
+	// on macOS, there's already buffer for available vram (see below) so just return the total
+	systemMemory := int64(memory.TotalMemory())
+
+	// macOS limits how much memory is available to the GPU based on the amount of system memory
+	// TODO: handle case where iogpu.wired_limit_mb is set to a higher value
+	if systemMemory <= 36*1024*1024*1024 {
+		systemMemory = systemMemory * 2 / 3
+	} else {
+		systemMemory = systemMemory * 3 / 4
+	}
+
+	return systemMemory, nil
 }
 
 func GetGPUInfo() GpuInfo {
-	// TODO - Metal vs. x86 macs...
 	mem, _ := getCPUMem()
 	return GpuInfo{
 		Library: "default",
@@ -30,19 +42,6 @@ func getCPUMem() (memInfo, error) {
 		TotalMemory: 0,
 		FreeMemory:  0,
 	}, nil
-}
-
-func NumGPU(numLayer, fileSizeBytes int64, opts api.Options) int {
-	if opts.NumGPU != -1 {
-		return opts.NumGPU
-	}
-
-	// metal only supported on arm64
-	if runtime.GOARCH == "arm64" {
-		return 1
-	}
-
-	return 0
 }
 
 func nativeInit() error {

--- a/gpu/gpu_darwin.go
+++ b/gpu/gpu_darwin.go
@@ -12,6 +12,7 @@ import (
 // CheckVRAM returns the free VRAM in bytes on Linux machines with NVIDIA GPUs
 func CheckVRAM() (int64, error) {
 	if runtime.GOARCH == "amd64" {
+		// gpu not supported, this may not be metal
 		return 0, nil
 	}
 

--- a/llm/ext_server_common.go
+++ b/llm/ext_server_common.go
@@ -35,14 +35,12 @@ import (
 	"encoding/json"
 	"fmt"
 	"log"
-	"os"
 	"strings"
 	"sync"
 	"time"
 	"unsafe"
 
 	"github.com/jmorganca/ollama/api"
-	"github.com/jmorganca/ollama/gpu"
 )
 
 type extServer interface {
@@ -82,25 +80,20 @@ func extServerResponseToErr(resp C.ext_server_resp_t) error {
 	return fmt.Errorf(C.GoString(resp.msg))
 }
 
-func newExtServer(server extServer, model string, adapters, projectors []string, numLayers int64, opts api.Options) (extServer, error) {
+func newExtServer(server extServer, model string, adapters, projectors []string, opts api.Options) (extServer, error) {
 	if !mutex.TryLock() {
 		log.Printf("concurrent llm servers not yet supported, waiting for prior server to complete")
 		mutex.Lock()
 	}
-	fileInfo, err := os.Stat(model)
-	if err != nil {
-		return nil, err
-	}
+
 	var sparams C.ext_server_params_t
 	sparams.model = C.CString(model)
 	defer C.free(unsafe.Pointer(sparams.model))
 
-	numGPU := gpu.NumGPU(numLayers, fileInfo.Size(), opts)
-
 	sparams.embedding = true
 	sparams.n_ctx = C.uint(opts.NumCtx)
 	sparams.n_batch = C.uint(opts.NumBatch)
-	sparams.n_gpu_layers = C.int(numGPU)
+	sparams.n_gpu_layers = C.int(opts.NumGPU)
 	sparams.main_gpu = C.int(opts.MainGPU)
 	sparams.n_parallel = 1 // TODO - wire up concurrency
 

--- a/llm/ext_server_default.go
+++ b/llm/ext_server_default.go
@@ -54,9 +54,9 @@ func (llm *llamaExtServer) llama_server_release_json_resp(json_resp **C.char) {
 	C.llama_server_release_json_resp(json_resp)
 }
 
-func newDefaultExtServer(model string, adapters, projectors []string, numLayers int64, opts api.Options) (extServer, error) {
+func newDefaultExtServer(model string, adapters, projectors []string, opts api.Options) (extServer, error) {
 	server := &llamaExtServer{opts}
-	return newExtServer(server, model, adapters, projectors, numLayers, opts)
+	return newExtServer(server, model, adapters, projectors, opts)
 }
 
 func (llm *llamaExtServer) Predict(ctx context.Context, pred PredictOpts, fn func(PredictResult)) error {

--- a/llm/ggml.go
+++ b/llm/ggml.go
@@ -78,7 +78,11 @@ type model interface {
 	ModelFamily() string
 	ModelType() string
 	FileType() string
-	NumLayers() int64
+	NumLayers() uint32
+	NumGQA() uint32
+	NumEmbed() uint32
+	NumHead() uint32
+	NumHeadKv() uint32
 }
 
 type container interface {

--- a/llm/gguf.go
+++ b/llm/gguf.go
@@ -272,14 +272,49 @@ func (llm *ggufModel) Decode(rso *readSeekOffset) error {
 	return nil
 }
 
-func (llm *ggufModel) NumLayers() int64 {
+func (llm *ggufModel) NumLayers() uint32 {
 	value, exists := llm.kv[fmt.Sprintf("%s.block_count", llm.ModelFamily())]
 	if !exists {
 		return 0
 	}
 
-	v := value.(uint32)
-	return int64(v)
+	return value.(uint32)
+}
+
+func (llm *ggufModel) NumHead() uint32 {
+	value, exists := llm.kv[fmt.Sprintf("%s.attention.head_count", llm.ModelFamily())]
+	if !exists {
+		return 0
+	}
+
+	return value.(uint32)
+}
+
+func (llm *ggufModel) NumEmbed() uint32 {
+	value, exists := llm.kv[fmt.Sprintf("%s.embedding_length", llm.ModelFamily())]
+	if !exists {
+		return 0
+	}
+
+	return value.(uint32)
+}
+
+func (llm *ggufModel) NumHeadKv() uint32 {
+	value, exists := llm.kv[fmt.Sprintf("%s.attention.head_count_kv", llm.ModelFamily())]
+	if !exists {
+		return 0
+	}
+
+	return value.(uint32)
+}
+
+func (llm *ggufModel) NumGQA() uint32 {
+	numHeadKv := llm.NumHeadKv()
+	if numHeadKv == 0 {
+		return 0
+	}
+
+	return llm.NumHead() / numHeadKv
 }
 
 func (llm ggufModel) readU8(r io.Reader) uint8 {

--- a/llm/llama.go
+++ b/llm/llama.go
@@ -8,7 +8,6 @@ import (
 	"fmt"
 	"os"
 	"os/exec"
-	"sync"
 	"time"
 
 	"github.com/jmorganca/ollama/api"
@@ -43,69 +42,11 @@ number ::= ("-"? ([0-9] | [1-9] [0-9]*)) ("." [0-9]+)? ([eE] [-+]? [0-9]+)? ws
 ws ::= ([ \t\n] ws)?
 `
 
-type llamaModel struct {
-	hyperparameters llamaHyperparameters
-}
-
-func (llm *llamaModel) ModelFamily() string {
-	return "llama"
-}
-
-func llamaModelType(numLayer uint32) string {
-	switch numLayer {
-	case 26:
-		return "3B"
-	case 32:
-		return "7B"
-	case 40:
-		return "13B"
-	case 48:
-		return "34B"
-	case 60:
-		return "30B"
-	case 80:
-		return "65B"
-	default:
-		return "unknown"
-	}
-}
-
-func (llm *llamaModel) ModelType() string {
-	return llamaModelType(llm.hyperparameters.NumLayer)
-}
-
-func (llm *llamaModel) FileType() string {
-	return fileType(llm.hyperparameters.FileType)
-}
-
-func (llm *llamaModel) NumLayers() int64 {
-	return int64(llm.hyperparameters.NumLayer)
-}
-
-type llamaHyperparameters struct {
-	// NumVocab is the size of the model's vocabulary.
-	NumVocab uint32
-
-	// NumEmbd is the size of the model's embedding layer.
-	NumEmbd uint32
-	NumMult uint32
-	NumHead uint32
-
-	// NumLayer is the number of layers in the model.
-	NumLayer uint32
-	NumRot   uint32
-
-	// FileType describes the quantization level of the model, e.g. Q4_0, Q5_K, etc.
-	FileType uint32
-}
-
 type Running struct {
 	Port          int
 	Cmd           *exec.Cmd
 	Cancel        context.CancelFunc
-	exitOnce      sync.Once
-	exitCh        chan error // channel to receive the exit status of the subprocess
-	*StatusWriter            // captures error messages from the llama runner process
+	*StatusWriter // captures error messages from the llama runner process
 }
 
 type ImageData struct {

--- a/llm/llm.go
+++ b/llm/llm.go
@@ -57,7 +57,7 @@ func New(workDir, model string, adapters, projectors []string, opts api.Options)
 	// TODO: use actual model size
 	requiredModel := ggml.Size
 
-	// fp16 k,v matrices each require = n_ctx * n_layer * n_embd / n_head * n_head_kv * 2 bytes each
+	// fp16 k,v matrices each require = n_ctx * n_layer * n_embd / n_head * n_head_kv * 2 bytes each * 2 key and value
 	requiredKv := 2 * 2 * int64(opts.NumCtx) * int64(ggml.NumLayers()) * int64(ggml.NumEmbed()) * int64(ggml.NumHeadKv()) / int64(ggml.NumHead())
 
 	// this amount is the overhead + tensors in memory

--- a/llm/llm.go
+++ b/llm/llm.go
@@ -7,10 +7,7 @@ import (
 	"os"
 	"runtime"
 
-	"github.com/pbnjay/memory"
-
 	"github.com/jmorganca/ollama/api"
-	"github.com/jmorganca/ollama/format"
 	"github.com/jmorganca/ollama/gpu"
 )
 
@@ -40,32 +37,74 @@ func New(workDir, model string, adapters, projectors []string, opts api.Options)
 		return nil, err
 	}
 
-	if runtime.GOOS == "darwin" {
-		var requiredMemory int64
-		var f16Multiplier int64 = 2
+	if opts.NumCtx < 4 {
+		opts.NumCtx = 4
+	}
 
-		switch ggml.ModelType() {
-		case "3B", "7B":
-			requiredMemory = 8 * format.GigaByte
-		case "13B":
-			requiredMemory = 16 * format.GigaByte
-		case "30B", "34B", "40B":
-			requiredMemory = 32 * format.GigaByte
-		case "47B":
-			requiredMemory = 48 * format.GigaByte
-		case "65B", "70B":
-			requiredMemory = 64 * format.GigaByte
-		case "180B":
-			requiredMemory = 128 * format.GigaByte
-			f16Multiplier = 4
-		}
+	fmt.Println("size", ggml.Size)
+	fmt.Println("filetype", ggml.FileType())
+	fmt.Println("architecture", ggml.ModelFamily())
+	fmt.Println("type", ggml.ModelType())
+	fmt.Println("name", ggml.Name())
+	fmt.Println("embd", ggml.NumEmbed())
+	fmt.Println("head", ggml.NumHead())
+	fmt.Println("head_kv", ggml.NumHeadKv())
+	fmt.Println("gqa", ggml.NumGQA())
 
-		systemMemory := int64(memory.TotalMemory())
+	available, _ := gpu.CheckVRAM()
 
-		if ggml.FileType() == "F16" && requiredMemory*f16Multiplier > systemMemory {
-			return nil, fmt.Errorf("F16 model requires at least %s of memory", format.HumanBytes(requiredMemory))
-		} else if requiredMemory > systemMemory {
-			return nil, fmt.Errorf("model requires at least %s of memory", format.HumanBytes(requiredMemory))
+	// For now assume filesize = model size
+	// TODO: use actual model size
+	requiredModel := ggml.Size
+
+	// fp16 k,v matrices each require = n_ctx * n_layer * n_embd / n_head * n_head_kv * 2 bytes each
+	requiredKv := 2 * 2 * int64(opts.NumCtx) * int64(ggml.NumLayers()) * int64(ggml.NumEmbed()) * int64(ggml.NumHeadKv()) / int64(ggml.NumHead())
+
+	// this amount is the overhead + tensors in memory
+	// TODO: get this from the llama.cpp's graph calcluations instead of
+	// guessing it's ~1/8th of the kv cache times gqa
+	// this is slightly (~25%) higher than what it is in reality
+	requiredAlloc := int64(ggml.NumGQA()) * requiredKv / 8
+
+	fmt.Println("system memory ", available)
+	fmt.Println("required model", requiredModel)
+	fmt.Println("required kv", requiredKv)
+	fmt.Println("required alloc", requiredAlloc)
+	fmt.Println("required total", requiredModel+requiredKv+requiredAlloc)
+
+	if opts.NumGPU == -1 {
+		opts.NumGPU = int(ggml.NumLayers())
+	}
+
+	// decide how many layers to put on the GPU
+	if opts.NumGPU > 0 {
+		switch runtime.GOOS {
+		case "darwin":
+			if requiredModel+requiredKv+requiredAlloc > available {
+				fmt.Println("not enough vram available, falling back to CPU only")
+				opts.NumGPU = 0
+			}
+		default:
+			info := gpu.GetGPUInfo()
+			if info.Library == "cpu" || info.Library == "default" {
+				opts.NumGPU = 0
+				break
+			}
+
+			// model and kv cache get loaded into vram
+			vram := requiredModel + requiredKv
+
+			var layers int64
+			if vram > available {
+				// TODO: calculate actual bytes per layer vs this estimation
+				bytesPerLayer := int64(vram / int64(ggml.NumLayers()))
+				layers = vram / bytesPerLayer
+			}
+
+			if layers < int64(opts.NumGPU) {
+				fmt.Println("only loading", layers, "layers")
+				opts.NumGPU = int(layers)
+			}
 		}
 	}
 
@@ -73,7 +112,7 @@ func New(workDir, model string, adapters, projectors []string, opts api.Options)
 	opts.RopeFrequencyBase = 0.0
 	opts.RopeFrequencyScale = 0.0
 	gpuInfo := gpu.GetGPUInfo()
-	return newLlmServer(gpuInfo.Library, model, adapters, projectors, ggml.NumLayers(), opts)
+	return newLlmServer(gpuInfo.Library, model, adapters, projectors, opts)
 }
 
 // Give any native cgo implementations an opportunity to initialize
@@ -81,9 +120,9 @@ func Init(workdir string) error {
 	return nativeInit(workdir)
 }
 
-func newLlmServer(library, model string, adapters, projectors []string, numLayers int64, opts api.Options) (extServer, error) {
+func newLlmServer(library, model string, adapters, projectors []string, opts api.Options) (extServer, error) {
 	if _, libPresent := AvailableShims[library]; libPresent && library != "default" {
-		srv, err := newDynamicShimExtServer(AvailableShims[library], model, adapters, projectors, numLayers, opts)
+		srv, err := newDynamicShimExtServer(AvailableShims[library], model, adapters, projectors, opts)
 		if err == nil {
 			return srv, nil
 		}
@@ -91,6 +130,5 @@ func newLlmServer(library, model string, adapters, projectors []string, numLayer
 		// TODO - update some state to indicate we were unable to load the GPU library for future "info" ux
 	}
 
-	return newDefaultExtServer(model, adapters, projectors, numLayers, opts)
-
+	return newDefaultExtServer(model, adapters, projectors, opts)
 }

--- a/llm/llm.go
+++ b/llm/llm.go
@@ -116,9 +116,6 @@ func New(workDir, model string, adapters, projectors []string, opts api.Options)
 			log.Println("splitting", available, "of available memory bytes into layers")
 			bytesPerLayer := int64((requiredModel + requiredKv) / int64(ggml.NumLayers()))
 			log.Println("bytes per layer:", bytesPerLayer)
-
-			// leave some room for overhead
-			// available = (available * 9) / 10
 			layers := available / bytesPerLayer
 			if layers < int64(opts.NumGPU) {
 				opts.NumGPU = int(layers)

--- a/llm/llm.go
+++ b/llm/llm.go
@@ -62,10 +62,8 @@ func New(workDir, model string, adapters, projectors []string, opts api.Options)
 
 	// this amount is the overhead + tensors in memory
 	// TODO: get this from the llama.cpp's graph calcluations instead of
-	// guessing it's ~1/8th of the kv cache times gqa
-	// this is slightly (~25%) higher than what it is in reality
-	// TODO: this needs to account for batch size != 512
-	requiredAlloc := int64(ggml.NumGQA()) * requiredKv / 8
+	// guessing it's ~1/7th of the kv cache times gqa
+	requiredAlloc := int64(ggml.NumGQA()) * requiredKv / 7
 
 	requiredTotal := requiredModel + requiredKv + requiredAlloc
 

--- a/llm/llm.go
+++ b/llm/llm.go
@@ -64,23 +64,26 @@ func New(workDir, model string, adapters, projectors []string, opts api.Options)
 	// TODO: get this from the llama.cpp's graph calcluations instead of
 	// guessing it's ~1/8th of the kv cache times gqa
 	// this is slightly (~25%) higher than what it is in reality
+	// TODO: this needs to account for batch size != 512
 	requiredAlloc := int64(ggml.NumGQA()) * requiredKv / 8
+
+	requiredTotal := requiredModel + requiredKv + requiredAlloc
 
 	fmt.Println("system memory ", available)
 	fmt.Println("required model", requiredModel)
 	fmt.Println("required kv", requiredKv)
 	fmt.Println("required alloc", requiredAlloc)
-	fmt.Println("required total", requiredModel+requiredKv+requiredAlloc)
+	fmt.Println("required total", requiredTotal)
 
 	if opts.NumGPU == -1 {
-		opts.NumGPU = int(ggml.NumLayers())
+		opts.NumGPU = int(ggml.NumLayers()) + 1
 	}
 
 	// decide how many layers to put on the GPU
 	if opts.NumGPU > 0 {
 		switch runtime.GOOS {
 		case "darwin":
-			if requiredModel+requiredKv+requiredAlloc > available {
+			if requiredTotal > available {
 				fmt.Println("not enough vram available, falling back to CPU only")
 				opts.NumGPU = 0
 			}
@@ -91,18 +94,27 @@ func New(workDir, model string, adapters, projectors []string, opts api.Options)
 				break
 			}
 
-			// model and kv cache get loaded into vram
-			vram := requiredModel + requiredKv
-
-			var layers int64
-			if vram > available {
-				// TODO: calculate actual bytes per layer vs this estimation
-				bytesPerLayer := int64(vram / int64(ggml.NumLayers()))
-				layers = vram / bytesPerLayer
+			if requiredTotal <= available {
+				break
 			}
 
+			// overhead + tensors are always loaded into
+			// scratch memory and so don't offload if there isn't even
+			// enough room for that
+			if requiredAlloc > available {
+				fmt.Println("not enough vram available, falling back to CPU only")
+				opts.NumGPU = 0
+				break
+			}
+
+			available -= requiredAlloc
+
+			// fill remaining vram with layers
+			fmt.Println("splitting", available, "into layers")
+			bytesPerLayer := int64((requiredModel + requiredKv) / int64(ggml.NumLayers()))
+			fmt.Println("bytes per layer:", bytesPerLayer)
+			layers := available/bytesPerLayer + 1
 			if layers < int64(opts.NumGPU) {
-				fmt.Println("only loading", layers, "layers")
 				opts.NumGPU = int(layers)
 			}
 		}

--- a/llm/llm.go
+++ b/llm/llm.go
@@ -77,6 +77,7 @@ func New(workDir, model string, adapters, projectors []string, opts api.Options)
 	library := info.Library
 
 	if opts.NumGPU == -1 {
+		// default to offloading all layers
 		opts.NumGPU = int(ggml.NumLayers()) + 1
 	}
 

--- a/llm/llm.go
+++ b/llm/llm.go
@@ -57,7 +57,7 @@ func New(workDir, model string, adapters, projectors []string, opts api.Options)
 	// TODO: use actual model size
 	requiredModel := ggml.Size
 
-	// fp16 k,v matrices each require = n_ctx * n_layer * n_embd / n_head * n_head_kv * 2 bytes each * 2 key and value
+	// fp16 k,v matrices require = n_ctx * n_layer * n_embd / n_head * n_head_kv * 2 bytes each * 2 key and value
 	requiredKv := 2 * 2 * int64(opts.NumCtx) * int64(ggml.NumLayers()) * int64(ggml.NumEmbed()) * int64(ggml.NumHeadKv()) / int64(ggml.NumHead())
 
 	// this amount is the overhead + tensors in memory

--- a/llm/llm.go
+++ b/llm/llm.go
@@ -69,11 +69,14 @@ func New(workDir, model string, adapters, projectors []string, opts api.Options)
 
 	requiredTotal := requiredModel + requiredKv + requiredAlloc
 
-	fmt.Println("system memory ", available)
-	fmt.Println("required model", requiredModel)
-	fmt.Println("required kv", requiredKv)
-	fmt.Println("required alloc", requiredAlloc)
-	fmt.Println("required total", requiredTotal)
+	log.Println("system memory bytes:", available)
+	log.Println("required model bytes:", requiredModel)
+	log.Println("required kv bytes:", requiredKv)
+	log.Println("required alloc bytes:", requiredAlloc)
+	log.Println("required total bytes:", requiredTotal)
+
+	info := gpu.GetGPUInfo()
+	library := info.Library
 
 	if opts.NumGPU == -1 {
 		opts.NumGPU = int(ggml.NumLayers()) + 1
@@ -84,25 +87,26 @@ func New(workDir, model string, adapters, projectors []string, opts api.Options)
 		switch runtime.GOOS {
 		case "darwin":
 			if requiredTotal > available {
-				fmt.Println("not enough vram available, falling back to CPU only")
+				log.Println("not enough vram available, falling back to CPU only")
 				opts.NumGPU = 0
 			}
 		default:
-			info := gpu.GetGPUInfo()
-			if info.Library == "cpu" || info.Library == "default" {
+			if library == "cpu" || library == "default" {
 				opts.NumGPU = 0
 				break
 			}
 
+			// no offloading required
 			if requiredTotal <= available {
 				break
 			}
 
-			// overhead + tensors are always loaded into
-			// scratch memory and so don't offload if there isn't even
-			// enough room for that
-			if requiredAlloc > available {
-				fmt.Println("not enough vram available, falling back to CPU only")
+			// This handles two cases:
+			// 1. overhead + tensors are always loaded into scratch memory even with num_gpu 0
+			// 2. it seems llama.cpp always tries to allocate the entire kv cache (even if later split into layers) into vram or crashes
+			if requiredAlloc > available || requiredKv > available {
+				log.Printf("not enough vram available, falling back to CPU only")
+				library = "cpu"
 				opts.NumGPU = 0
 				break
 			}
@@ -110,10 +114,13 @@ func New(workDir, model string, adapters, projectors []string, opts api.Options)
 			available -= requiredAlloc
 
 			// fill remaining vram with layers
-			fmt.Println("splitting", available, "into layers")
+			log.Println("splitting", available, "of available memory bytes into layers")
 			bytesPerLayer := int64((requiredModel + requiredKv) / int64(ggml.NumLayers()))
-			fmt.Println("bytes per layer:", bytesPerLayer)
-			layers := available/bytesPerLayer + 1
+			log.Println("bytes per layer:", bytesPerLayer)
+
+			// leave some room for overhead
+			// available = (available * 9) / 10
+			layers := available / bytesPerLayer
 			if layers < int64(opts.NumGPU) {
 				opts.NumGPU = int(layers)
 			}

--- a/llm/shim_darwin.go
+++ b/llm/shim_darwin.go
@@ -16,7 +16,7 @@ import (
 //go:embed llama.cpp/ggml-metal.metal
 var libEmbed embed.FS
 
-func newDynamicShimExtServer(library, model string, adapters, projectors []string, numLayers int64, opts api.Options) (extServer, error) {
+func newDynamicShimExtServer(library, model string, adapters, projectors []string, opts api.Options) (extServer, error) {
 	// should never happen...
 	return nil, fmt.Errorf("Dynamic library loading not supported on Mac")
 }

--- a/llm/shim_ext_server.go
+++ b/llm/shim_ext_server.go
@@ -72,7 +72,7 @@ func (llm *shimExtServer) llama_server_release_json_resp(json_resp **C.char) {
 	C.dynamic_shim_llama_server_release_json_resp(llm.s, json_resp)
 }
 
-func newDynamicShimExtServer(library, model string, adapters, projectors []string, numLayers int64, opts api.Options) (extServer, error) {
+func newDynamicShimExtServer(library, model string, adapters, projectors []string, opts api.Options) (extServer, error) {
 	shimMutex.Lock()
 	defer shimMutex.Unlock()
 	updatePath(filepath.Dir(library))
@@ -90,7 +90,7 @@ func newDynamicShimExtServer(library, model string, adapters, projectors []strin
 		options: opts,
 	}
 	log.Printf("Loading Dynamic Shim llm server: %s", library)
-	return newExtServer(llm, model, adapters, projectors, numLayers, opts)
+	return newExtServer(llm, model, adapters, projectors, opts)
 }
 
 func (llm *shimExtServer) Predict(ctx context.Context, pred PredictOpts, fn func(PredictResult)) error {

